src/compiler/c.rs

```diff
@@ -383,21 +383,23 @@ where
         // request.
         let preprocessor_cache_mode_config = storage.preprocessor_cache_mode_config();
         // Disable preprocessor cache when doing distributed compilation
-        let mut preprocessor_key =
-            if !may_dist && preprocessor_cache_mode_config.use_preprocessor_cache_mode {
-                preprocessor_cache_entry_hash_key(
-                    &executable_digest,
-                    parsed_args.language,
-                    &preprocessor_and_arch_args,
-                    &extra_hashes,
-                    &env_vars,
-                    &absolute_input_path,
-                    compiler.plusplus(),
-                    preprocessor_cache_mode_config,
-                )?
-            } else {
-                None
-            };
+        let mut preprocessor_key = if !may_dist
+            && preprocessor_cache_mode_config.use_preprocessor_cache_mode
+            && !parsed_args.too_hard_for_preprocessor_cache_mode
+        {
+            preprocessor_cache_entry_hash_key(
+                &executable_digest,
+                parsed_args.language,
+                &preprocessor_and_arch_args,
+                &extra_hashes,
+                &env_vars,
+                &absolute_input_path,
+                compiler.plusplus(),
+                preprocessor_cache_mode_config,
+            )?
+        } else {
+            None
```

Comment:
That would need to mention what specifically makes it too hard for the preprocessor cache, though, otherwise the log is not going to be useful (like `cannot_cache!`). Maybe more of a followup?

Reply:
Done in #2197. I'll rebase this one when that one is merged.